File: 0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        para = {"{":"}", "[":"]", "(":")"}
        for x in s:
            if x in para:
                stack.append(x)
            else:
                if not stack:
                    return False
                temp = stack.pop()
                if para[temp] != x:
                    return False
        if not stack:
            return True
        else:
            return False

File: 0020-valid-parentheses/solution.cpp
using namespace std;
#include <vector>
#include <string>
#include <stack>
#include <unordered_map>

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> para = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        stack<char> stack = {};
        for(char cur:s){
            if (para.count(cur)){
                stack.push(cur);
            }
            else{
                if(stack.empty()){
                    return false;
                }
                if (para[stack.top()] == cur){
                    stack.pop();
                }
                else{
                    return false;
                }
            }

        }

        if(stack.empty()){
            return true;
        }
        return false;
    }
};
